File: Data Structures & Algorithms/spiral-matrix/submission-0.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        int m = matrix.size(), n = matrix[0].size();
        int top = 0, bottom = m-1, left = 0, right = n-1;

        vector<int> res;

        int r, c;
        while(top <= bottom && left <= right){

        r = top;

        for(c = left; c<=right; c++){
        matrix[r][c];
        res.push_back( matrix[r][c]);
        }
        top++;

        c = right;

        for(r = top; r<=bottom; r++){
        matrix[r][c];
        res.push_back( matrix[r][c]);
        }
        right--;


        if(top <= bottom){
        r = bottom;
        for(c = right; c>=left; c--){
        matrix[r][c];
        res.push_back( matrix[r][c]);
        }
        bottom--;
        }

        if(left <= right){
        c = left;
        for(r = bottom; r>=top; r--){
        matrix[r][c];
        res.push_back( matrix[r][c]);
        }
        left++;
        }
        }
        return res;
    }
};

File: Data Structures & Algorithms/pacific-atlantic-water-flow/submission-0.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        m = len(heights)
        n = len(heights[0])

        pacific = set()
        atlantic = set()

        def dfs(r, c, visited):
            visited.add((r, c))

            directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]

            for dr, dc in directions:
                nr = r + dr
                nc = c + dc

                if (
                    0 <= nr < m
                    and 0 <= nc < n
                    and (nr, nc) not in visited
                    and heights[nr][nc] >= heights[r][c]
                ):
                    dfs(nr, nc, visited)

        # Pacific: top row
        for c in range(n):
            dfs(0, c, pacific)

        # Pacific: left column
        for r in range(m):
            dfs(r, 0, pacific)

        # Atlantic: bottom row
        for c in range(n):
            dfs(m - 1, c, atlantic)

        # Atlantic: right column
        for r in range(m):
            dfs(r, n - 1, atlantic)

        # Find cells reachable from both oceans
        result = []

        for r in range(m):
            for c in range(n):
                if (r, c) in pacific and (r, c) in atlantic:
                    result.append([r, c])

        return result